### ISSUE TITLE
Variable leak via unification

### ISSUE THREAD

**Issue by liesnikov** [liesnikov/extensible-elaborator#6]
As of 1641acaa276c3f when typechecking `pi/test/CaseTest.pi` we're leaking variable `b` in the case construct.

This is because we're unifying in the context that carries still a free named variable, but the substitution of the solution happens into the term where this variable has already been captured by the lambda.

some (unsatisfactory) solutions

* switch away from locally nameless (and unbound)
* only bind if we know that the body doesn't contain any metas and otherwise put a metavariable placeholder
* track when a variable gets bound and use it instead in the context. This seems like the hackiest solution and I'm not sure about the implications yet.

**Comment by liesnikov** [comment]
The error is 
```
type checking...
Checking module "CaseTest"
Type Error:
The variable b17 was not found.
in context
  b : Bool
  not : Bool -> Bool
When checking the term 
  \b.
    case b of
      True -> False
      False -> True
against the signature
  not : Bool -> Bool
```

The b in the context is indeed distinct, it's `b1` and not `b17`.

**Comment by liesnikov** [comment]
Solved by implementing closures

**Comment by liesnikov** [comment]
fixed in c491298
